src/main/java/org/javarosa/core/services/storage/IStorageUtilityIndexed.java

```diff
@@ -200,11 +201,24 @@
      */
     E getRecordForValue(String fieldName, Object value) throws NoSuchElementException, InvalidIndexException;
 
+    void bulkRead(LinkedHashSet<Integer> ids, HashMap<Integer, E> recordMap);
+
     /**
      * Load multiple record objects from storage at one time from a list of record ids.
-     *
+     * <p>
      * If the provided recordMap already contains entries for any ids, it is _not_
      * required for them to be retrieved from storage again.
+     *
+     * Since this method can have a significant runtime, an abandonable is provided to enable the
+     * request to be shortcircuited. Implementations should regularly assert that the request
+     * has not been abandoned on long-running bulk reads
+     * 
+     * @throws RequestAbandonedException If the current request is abandoned, this method will
+     *                                   throw a RequestAbandonedException. Callers should not
+     *                                   generally catch that exception unless they rethrow it
+     *                                   or another exception, but they should anticipate that
+     *                                   they may need to clean up if the bulk read doesn't complete
      */
-    void bulkRead(LinkedHashSet<Integer> ids, HashMap<Integer, E> recordMap);
-}
\ No newline at end of file
+
+    void bulkRead(LinkedHashSet<Integer> cuedCases, HashMap<Integer, E> recordMap, Abandonable abandonable) throws RequestAbandonedException;
```

Comment:
@ctsims The build is failing b/c this needs an import statement for `RequestAbandonedException`. And shouldn't the overriding methods have this added to their declaration as well?

Reply:
Hey, yeay, sorry, not quite done putting in those changes, sorry, pushed early.